pkg/discovery/worker/controller_discovery_worker.go

```diff
@@ -0,0 +1,80 @@
+package worker
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/dtofactory"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/metrics"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+)
+
+// Converts the cluster K8s controller objects to entity DTOs
+type K8sControllerDiscoveryWorker struct {
+	cluster *repository.ClusterSummary
+}
+
+func NewK8sControllerDiscoveryWorker(cluster *repository.ClusterSummary) *K8sControllerDiscoveryWorker {
+	return &K8sControllerDiscoveryWorker{
+		cluster: cluster,
+	}
+}
+
+// Controller discovery worker collects KubeController entities discovered by different discovery workers.
+// It merges the pods belonging to the same controller but discovered by different discovery workers, and aggregates
+// allocation resources usage of the same KubeController from different workers.
+// Then it creates entity DTOs for the K8s controllers to be sent to the Turbonomic server.
+func (worker *K8sControllerDiscoveryWorker) Do(kubeControllers []*repository.KubeController) ([]*proto.EntityDTO, error) {
+	namespacesMap := worker.cluster.NamespaceMap
+	// Map from controller UID to the corresponding kubeController
+	kubeControllersMap := make(map[string]*repository.KubeController)
+	// Combine the pods and allocation resources usage from different discovery workers but belonging to the same K8s controller.
+	for _, kubeController := range kubeControllers {
+		existingKubeController, exists := kubeControllersMap[kubeController.UID]
+		if !exists {
+			kubeNamespace, exists := namespacesMap[kubeController.Namespace]
+			if !exists {
+				glog.Errorf("Namespace %s does not exist in cluster %s", kubeController.Namespace, worker.cluster.Name)
+				continue
+			}
+			for resourceType, resource := range kubeController.AllocationResources {
+				// For CPU resources, convert the capacity values expressed in number of cores to MHz.
+				// Skip the conversion if capacity value is repository.DEFAULT_METRIC_CAPACITY_VALUE (infinity), which
+				// means resource quota is not configured on the corresponding namespace.
+				if metrics.IsCPUType(resourceType) && resource.Capacity != repository.DEFAULT_METRIC_CAPACITY_VALUE &&
+					kubeNamespace.AverageNodeCpuFrequency > 0.0 {
```

Comment:
Under what circumstances would averageNodeCpuFrequency not be greater than zero? 

Since this value doesn't change inside the loop, would it make sense to move this check outside the loop and then log a warning if it is <= 0 instead of just silently skipping the conversion?